### PR TITLE
Add DSA over GQA to the hybrid model

### PR DESCRIPTION
**Pull request by alokpathy**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
This is the second of four stacked PRs that introduces a simplified version of DSA. These stacked PRs split PR #[6869](https://github.com/NVIDIA/Megatron-LM/pull/6869).

**Indexer + Training** This PR adds simplified DSA over GQA: a learned indexer selects the top-k keys each query attends to, and main attention runs over only that subset of keys. A summary of the changes in each file is included below.

**Wiring**: This PR also adds the necessary wiring. `DSGroupedSelfAttention` subclasses `SelfAttention` and replaces `submodules.core_attention` with `DSGQACoreAttention`. The original dense core attention is passed through as `dense_core_attention` and instantiated only when `dsa_fwd_skip_dsa` or `dsa_fwd_use_dense_attn` is set — the former runs plain dense attention with a zero indexer loss, the latter runs dense attention for the output while still training the indexer against it. Both are necessary for each stage of DSA training.

`attention.py` gains one hook, `_get_core_attention_extra_kwargs`, returning `{}` by default. DSA needs `hidden_states` inside core attention because the indexer projects from it rather than from Q/K/V.

| File | Change |
|---|---|
| `megatron/core/transformer/experimental_attention_variant/dsa_gqa.py` | **New.** Simplified indexer over GQA: `DSGroupedSelfAttention`, `DSGQACoreAttention`, the simplified indexer, the PyTorch reference sparse-attention paths (dense-mask and gather), and the indexer KL loss. |
| `megatron/core/transformer/experimental_attention_variant/dsa_layer_specs.py` | **New.** `dsa_stack_spec`, derived from upstream's `hybrid_stack_spec`. |
| `megatron/core/transformer/experimental_attention_variant/dsa.py` | Query and key chunked top-k helpers (`fused_qk_topk_chunked`, `_merge_topk_scores`), and indexer loss |
| `megatron/core/transformer/attention.py` | Two base-class hooks: `_get_core_attention_extra_kwargs` (pass extra inputs into core attention) and `_dynamic_core_attention_forward` (raises by default so DSA + dynamic batching fails loudly instead of silently running dense). Needed to pass indexer parameters into Attention forward. |
| `megatron/core/transformer/transformer_config.py` | 31 `dsa_*` fields and validation. Adds six rope fields to base config because the GQA indexer builds its own `RotaryEmbedding` from `config`. |

## Part of a series

This PR is 2 of 4 splitting a single 13k-line PR #6869 

| PR | Contents |
|----|----------|
| #7152     | min-memory DSA kernels (torch + Triton) |
| **this PR**   | DSA-over-GQA attention layer            |
| #7158   | per-parameter-group optimizer clocks    |
| #7181    | indexer freezing, reset, checkpointing  |

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/transformer/transformer_config.py [hunk 1]
```diff
@@ -343,10 +373,71 @@ class TransformerConfig(ModelParallelConfig):
...
+    dsa_min_memory_profile: bool = False
+    """Whether to print per-layer DSA min-memory forward/backward timing breakdowns."""
+
+    dsa_min_memory_profile_rank: int = 0
+    """Global rank that prints DSA min-memory timings. Set to -1 to print on every rank."""
+
+    dsa_kernel_query_block_size: Optional[int] = None
+    """Optional query tile size for DSA min-memory kernel backends."""
+
+    dsa_kernel_key_block_size: Optional[int] = None
```
Comment (zhongbozhu) [inline megatron/core/transformer/transformer_config.py:367]: I don't think it's appropriate to expose kernel tile size.

**Review by zhongbozhu — CHANGES_REQUESTED** [review]

File: megatron/core/transformer/transformer_config.py [hunk 2]
```diff
@@ -370,6 +461,15 @@ class TransformerConfig(ModelParallelConfig):
 
     dsa_indexer_k_norm_fp32: bool = False
     """Whether DSA indexer key LayerNorm should run on fp32 inputs."""
+    dsa_indexer_sparse_loss_use_topk_only: bool = False
+    """When using sparse DSA indexer loss, compute KL only on the selected top-k support."""
+
+    dsa_indexer_use_hadamard: bool = False
+    """Whether to apply Hadamard rotation to DSA indexer queries and keys."""
+
+    dsa_use_cudnn: bool = False
```
Comment (zhongbozhu) [inline megatron/core/transformer/transformer_config.py]: this duplicates ` dsa_kernel_backend: Literal["none", "tilelang", "cudnn"] = "none"`

**Review by zhongbozhu — CHANGES_REQUESTED** [review]

File: megatron/core/transformer/transformer_config.py [hunk 3]
```diff
@@ -258,6 +258,15 @@ class TransformerConfig(ModelParallelConfig):
...
 
+    rotary_seq_len_interpolation_factor: Optional[float] = None
+    """Sequence-length interpolation factor used by the model and DSA indexer RoPE."""
+
+    use_rope_scaling: bool = False
+    """Whether to use Llama-style RoPE frequency scaling."""
+
+    rope_scaling_factor: float = 8.0
+    """Scaling factor used when ``use_rope_scaling`` is enabled."""
+
```
Comment (Mellonta) [inline megatron/core/transformer/transformer_config.py]: Are they duplicated? defined in `megatron/training/arguments.py` already

File: megatron/core/transformer/transformer_config.py [hunk 4]
```diff
@@ -309,6 +318,15 @@ class TransformerConfig(ModelParallelConfig):
...
 
+    rope_type: str = "rope"
+    """Type of RoPE to use. Common attention defaults to rope."""
+
+    rotary_base: float = 10000
+    """Rotary base for the rotary embeddings."""
+
+    rotary_percent: float = 1.0
+    """Rotary percent for the rotary embeddings."""
+
```
Comment (Mellonta) [inline megatron/core/transformer/transformer_config.py]: same as above

File: megatron/core/transformer/transformer_config.py
Comment (Phlip79) [inline megatron/core/transformer/transformer_config.py:1]: Are all these knobs actually used? This seems excessive. I also see a lot of duplication.

File: megatron/core/transformer/transformer_config.py [hunk 5]
```diff
@@ -3410,8 +3510,269 @@ def _scope_to_str(s):
             assert not self.add_qkv_bias
             assert not self.use_kitchen
 
+        assert (
```
Comment (Phlip79) [inline megatron/core/transformer/transformer_config.py:3501]: This is hard to read and understand. Can it be simplified? Is it necessary?

File: megatron/core/transformer/transformer_config.py (hunk 1 above)
Comment (alokpathy) [reply megatron/core/transformer/transformer_config.py]: Removed these flags in favor of a memory budget.

File: megatron/core/transformer/transformer_config.py (hunk 2 above)
Comment (alokpathy) [reply megatron/core/transformer/transformer_config.py]: Folded this into `dsa_kernel_backend`

File: megatron/core/transformer/transformer_config.py (hunk 3 above)
Comment (alokpathy) [reply megatron/core/transformer/transformer_config.py]: Removed, thanks!

File: megatron/core/transformer/transformer_config.py (hunk 4 above)
Comment (alokpathy) [reply megatron/core/transformer/transformer_config.py]: Removed, thanks!

File: megatron/core/transformer/transformer_config.py
Comment (alokpathy) [reply megatron/core/transformer/transformer_config.py:1]: Thanks! Not all were necessary. I removed the ones that were the most disposable.

File: megatron/core/transformer/transformer_config.py (hunk 5 above)
Comment (alokpathy) [reply megatron/core/transformer/transformer_config.py]: Simplified, thanks!

File: megatron/core/transformer/experimental_attention_variant/dsa_gqa.py [hunk 6]
```diff
@@ -0,0 +1,1168 @@
...
+                        query_detached,
+                        key_detached,
+                        self.softmax_scale,
+                        indexer_loss_coeff,
+                        sparse_indexer_loss,
+                        self.indexer.pg_collection,
+                    )
+
+                indexer_loss = _compute_indexer_loss(index_scores)
+                DSAIndexerLossLoggingHelper.save_loss_to_tracker(
```
Comment (Mellonta) [inline megatron/core/transformer/experimental_attention_variant/dsa_gqa.py:856]: `save_loss_to_tracker()` doesn’t have `raw_loss` argument? The same for other calls

File: megatron/core/transformer/experimental_attention_variant/dsa_gqa.py [hunk 7]
```diff
@@ -0,0 +1,1168 @@
...
+        if not simplified_indexer and not getattr(
+            self.config, "dsa_indexer_rotate_activation", True
+        ):
+            raise NotImplementedError(
+                f"dsa_kernel_backend='{dsa_kernel_backend}' requires "
+                "dsa_indexer_rotate_activation."
+            )
+        if (
+            self.config.fp8 is not None
+            or self.config.fp8_param
```
Comment (Mellonta) [inline megatron/core/transformer/experimental_attention_variant/dsa_gqa.py:947]: Let's use get_fp8_disabled_context around indexer init/forward to relax this check.

File: megatron/core/transformer/experimental_attention_variant/dsa_gqa.py [hunk 8]
```diff
@@ -0,0 +1,1168 @@
...
+                    original_max_position_embeddings=config.original_max_position_embeddings,
+                    beta_fast=config.beta_fast,
+                    beta_slow=config.beta_slow,
+                    mscale=config.mscale,
+                    mscale_all_dim=config.mscale_all_dim,
+                    use_cpu_initialization=config.use_cpu_initialization,
+                    cp_group=self.pg_collection.cp,
+                )
+
+        self.linear_q = build_module(
```
Comment (Mellonta) [inline megatron/core/transformer/experimental_attention_variant/dsa_gqa.py:524]: +        with get_fp8_disabled_context(config, is_init=True):
same for self.linear_k

File: megatron/core/transformer/experimental_attention_variant/dsa_gqa.py [hunk 9]
```diff
@@ -0,0 +1,1168 @@
...
+        use_rope: bool,
+        packed_seq_params: Optional[PackedSeqParams] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Project the indexer's Q and its separately learned K from the layer input."""
+        if self.config.sequence_parallel and self.pg_collection.tp.size() > 1:
+            hidden_states = gather_from_sequence_parallel_region(
+                hidden_states, group=self.pg_collection.tp
+            )
+        seqlen, batch_size, _ = hidden_states.shape
+        q, _ = self.linear_q(hidden_states)
```
Comment (Mellonta) [inline megatron/core/transformer/experimental_attention_variant/dsa_gqa.py:586]: Also with get_fp8_disabled_context(self.config) here

File: megatron/core/transformer/experimental_attention_variant/dsa_layer_specs.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    Returns:
+        A deep copy with ``attention_layer``'s self-attention module replaced by
+        :class:`DSGroupedSelfAttention`.
+    """
+    spec = copy.deepcopy(stack_spec)
+    spec.submodules.attention_layer.submodules.self_attention.module = DSGroupedSelfAttention
+    return spec
+
+
+dsa_stack_spec = _with_dsa_gqa_attention(hybrid_stack_spec)
```
Comment (Mellonta) [inline megatron/core/transformer/experimental_attention_variant/dsa_layer_specs.py:43]: Let's add a 
```
gdp_dsa_stack_spec = _with_dsa_gqa_attention(gated_delta_product_stack_spec)
```

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 11]
```diff
@@ -420,6 +449,19 @@ def track_indexer_metrics(
             else:
                 total_loss_dict["indexer loss"] = avg_indexer_loss
 
+        if "raw_values" in tracker:
+            raw_values = tracker["raw_values"] * loss_scale
+            avg_raw_indexer_loss = raw_values.sum() / max(num_indexer_layers, 1)
```
Comment (Mellonta) [inline megatron/core/transformer/experimental_attention_variant/dsa.py:454]: It's now counting all layers, can we pass number of dsa layers?

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 12]
```diff
@@ -358,19 +380,26 @@ def reduce_loss_in_tracker(
...
+            # Reduce indexer losses across ranks.
+            if tracker.get('reduce_group') is not None:
+                torch.distributed.all_reduce(values, group=tracker.get('reduce_group'))
+            if tracker.get('avg_group') is not None:
+                torch.distributed.all_reduce(
+                    values, group=tracker['avg_group'], op=torch.distributed.ReduceOp.AVG
+                )
             torch.distributed.all_reduce(
-                values, group=tracker['avg_group'], op=torch.distributed.ReduceOp.AVG
+                values, group=pg_collection.dp, op=torch.distributed.ReduceOp.AVG
```
Comment (Mellonta) [inline megatron/core/transformer/experimental_attention_variant/dsa.py:401]: We also need to average across gtp group as well.

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 13]
```diff
@@ -2178,6 +2340,7 @@ def compute_indexer_loss_with_reference_path():
                 query_valid_rows,
                 self.config.calculate_per_token_loss,
                 self.config.dsa_indexer_scoring_relu,
+                getattr(self.config, 'dsa_kernel_backend', 'none') == 'cudnn',
```
Comment (Mellonta) [inline megatron/core/transformer/experimental_attention_variant/dsa.py:2343]: Missing a non_compressed_lse here?

File: megatron/core/transformer/transformer_config.py [hunk 14]
```diff
@@ -3410,8 +3455,145 @@ def _scope_to_str(s):
...
+            assert (
+                not dense_dsa_warmup or min_memory_dsa_backend
+            ), "dsa_fwd_use_dense_attn requires a min-memory dsa_kernel_backend."
+            # The simplified indexer has only ever been exercised on the GQA path. Rather than
+            # let an untested combination run, refuse it; the MLA path keeps the standard indexer.
+            assert not (
+                self.multi_latent_attention and self.dsa_indexer_mode == 'simplified'
+            ), "dsa_indexer_mode='simplified' is not supported with multi_latent_attention."
+
+            if not self.multi_latent_attention:
```
Comment (Mellonta) [inline megatron/core/transformer/transformer_config.py:3549]: Can we scope the simplified-indexer requirement to the new DSGroupedSelfAttention? The existing GQA adapter in tests/unit_tests/resharding/test_model_swap.py uses DSAttention with the standard indexer, so this global assertion rejects its configuration.

**Review by Mellonta — CHANGES_REQUESTED** [review]
